Add c8y_configuration_plugin how-to guide (#1151)
thin-edge/thin-edge.io: docs/src/howto-guides/025_config_management_plugin.md

# How to manage configuration files with Cumulocity

With `thin-edge.io`, you can manage config files on a device by using the [Cumulocity configuration management feature](https://cumulocity.com/guides/users-guide/device-management/#managing-configurations) as a part of Device Management.

If you are new to the Cumulocity **Configuration Management** feature,
we recommend you to read [the Cumulocity user guide](https://cumulocity.com/guides/users-guide/device-management/#managing-configurations) along with this how-to guide.

## Installation of `c8y_configuration_plugin`

To enable the feature, first you need to install the `c8y_configuration_plugin` binary on your device.

> ### Using the `get-thin-edge_io.sh` script for Debian OS distribution (Recommended)

Reviewer: To cover other Debian based distributions like Ubuntu as well.
```suggestion
### Using the `get-thin-edge_io.sh` script on Debian based distributions (Recommended)
```

If your device supports `apt` as a package manager,
you can install all `thin-edge.io` packages including the `c8y_configuration_plugin` by the `get-thin-edge_io.sh` script.
If you have already used the `get-thin-edge_io.sh` script,
this package is installed, by default.

```shell
curl -fsSL https://raw.githubusercontent.com/thin-edge/thin-edge.io/main/get-thin-edge_io.sh | sudo sh -s
```

> ### Using the `c8y_configuration_plugin` Debian package for Debian OS distribution

Reviewer: ```suggestion
### Using the `c8y_configuration_plugin` Debian package on Debian based distributions
```

> For Debian OS distribution, we provide the `c8y_configuration_plugin_<version>_<arch>.deb` package as a release asset [here](https://github.com/thin-edge/thin-edge.io/releases).

Reviewer: ```suggestion
For Debian based distributions, we provide the `c8y_configuration_plugin_<version>_<arch>.deb` package as a release asset [here](https://github.com/thin-edge/thin-edge.io/releases).
```

In case that you didn't use the `get-thin-edge_io.sh` script, you can download the `c8y_configuration_plugin_<version>_<arch>.deb`  package on our [Releases](https://github.com/thin-edge/thin-edge.io/releases) and install it.

```shell
sudo apt install ./path/to/package/c8y_configuration_plugin_<version>_<arch>.deb
```

> ### Extracting from debian package for non-Debian OS distribution

Reviewer: ```suggestion
### Extracting from debian package on non-Debian based distributions
```

For more details, refer to our guide [Extracting from debian package](./015_installation_without_deb_support.md#extracting-binaries-from-deb-packages).

### Building from sources

Follow our guide [Buiding thin-edge.io](./../BUILDING.md) and [Building from source](./015_installation_without_deb_support.md#if-building-from-source).

```shell
cargo build --release -p c8y_configuration_plugin
```

A `systemd` unit file for `c8y_configuration_plugin` can be found in the repository at `configuration/init/systemd/c8y-configuration-plugin.service`
and should be installed on the target in: `/lib/systemd/system/c8y-configuration-plugin.service`.

```shell
sudo cp <repository_root>/configuration/init/systemd/c8y-configuration-plugin.service /lib/systemd/system/c8y-configuration-plugin.service
```

## Get started

Before starting anything, make sure [your device is connected to Cumulocity](./../tutorials/connect-c8y.md).

**Step 0**
Only if you didn't install `c8y_configuration_plugin` by the debian package,
you need one additional step to initialize the plugin. Run this command.

```shell
sudo c8y_configuration_plugin --init
```

**Step 1**

> Open the file `/etc/tedge/c8y/c8y-configuration-plugin.toml` and edit the file.

Reviewer: ```suggestion
Open the file `/etc/tedge/c8y/c8y-configuration-plugin.toml` and add entries for the configuration files that you'd like this plugin to manage in the following format:
```

```toml
files = [

>     { path = '/etc/tedge/tedge.toml' },

Reviewer: ```suggestion
    { path = '/etc/tedge/tedge.toml', type = 'tedge.toml'},
```

Let's add a `type` for every entry and **NOT** document that it's an optional key for now, as we might have to make it mandatory in an upcoming release to ease the `tedge/configuration_change` MQTT notification mechanism.

Author: Sure

    { path = '/etc/tedge/mosquitto-conf/c8y-bridge.conf', type = 'c8y-bridge.conf' },
    { path = '/etc/tedge/mosquitto-conf/tedge-mosquitto.conf', type = 'tedge-mosquitto.conf' },
    { path = '/etc/mosquitto/mosquitto.conf', type = 'mosquitto.conf' },
    { path = '/etc/tedge/c8y/example.txt', type = 'example', user = 'tedge', group = 'tedge', mode = 0o444 }
]
```

> Add the files you want to configure from the cloud using the [format described in the reference guide](./../references/c8y-configuration-management.md#configuration).

Reviewer: ```suggestion
* `path` is the full path to the configuration file.
* `type` is a unique alias for each file entry which will be used to represent that file in Cumulocity UI.
* `user`, `group` and `mode` are UNIX file permission settings to be used to create a configuration file. If not provided, the files will be created with `root` user. If the file exists already, its ownership will be retained.

For more details on this configuration file format, refer to the [reference guide](./../references/c8y-configuration-management.md#configuration).
```

Reviewer 2: Indeed, easier for the user to have the key points inline and a link for the details,

## Update `c8y-configuration-plugin` from Cumulocity

To update any configuration file, create a local copy of that config file and then upload that file to the [Cumulocity configuration repository](https://cumulocity.com/guides/users-guide/device-management/#to-add-a-configuration-snapshot) with the appropriate configuration type.

The `c8y-configuration-plugin.toml` file can also be updated from the cloud in a similar manner to add/remove further configuration file entries. The updated TOML file has to be uploaded with the configuration type:  **c8y-configuration-plugin**.

Then, go back to the **Configuration** tab of your desired device in Cumulocity.

![Cumulocity Configuration Management Donwload](./images/c8y-config-plugin-download.png)

Click on the config file entry from the **DEVICE SUPPORTED CONFIGURATIONS** files list.
You can choose the file that you uploaded from the **AVAILABLE SUPPORTED CONFIGURATIONS** section, and then apply that file to your device by clicking on the **Send configuration to device** button.

After the operation created gets marked SUCCESSFUL, reload the page.
Then you can find new supported configuration types as you defined.

> 

Reviewer: We must document that `tedge/configuration_change` topic contract as well.

# How to manage configuration files with Cumulocity

With `thin-edge.io`, you can manage config files on a device by using the [Cumulocity configuration management feature](https://cumulocity.com/guides/users-guide/device-management/#managing-configurations) as a part of Device Management.

If you are new to the Cumulocity **Configuration Management** feature,
we recommend you to read [the Cumulocity user guide](https://cumulocity.com/guides/users-guide/device-management/#managing-configurations) along with this how-to guide.

## Installation of `c8y_configuration_plugin`

To enable the feature, first you need to install the `c8y_configuration_plugin` binary on your device.

### With debian package

For Debian distribution OS, we provide the `c8y_configuration_plugin_<version>_<arch>.deb` package as a release asset [here](https://github.com/thin-edge/thin-edge.io/releases).

If you have used `get_thin_edge_io.sh` script to install `thin-edge.io`, this package is installed, by default.

> You can skip this section.

Reviewer 2: Instead of having this *inside* the Debian section, I would add a 3rd sections. And even a  4th to separate the from source section.

* Install script (it requires `apt`)
* Using the Debian package
* For non-Debian OS distribution
* From source

Reviewer 2: I would also add the command lines for each case.

In case that you didn't use the script, you can get the package on our [Releases](https://github.com/thin-edge/thin-edge.io/releases) and install it.

### For non-Debian OS distribution

To get the `c8y_configuration_plugin` binary, you can either extract it from the debian package or build from the sources.

#### Extracting from debian package

Get the `c8y_configuration_plugin_<version>_<arch>.deb` from our [Releases](https://github.com/thin-edge/thin-edge.io/releases).
Then, follow our guide [Extracting from debian package](./015_installation_without_deb_support.md#extracting-binaries-from-deb-packages).

#### Building from sources

Follow our guide [Building from source](./015_installation_without_deb_support.md#if-building-from-source).

A `systemd` unit file for `c8y_configuration_plugin` can be found in the repository at `configuration/init/systemd/c8y-configuration-plugin.service`
and should be installed on the target in: `/lib/systemd/system/c8y-configuration-plugin.service`.

## Get started

> Before starting anything, first finish [establishing the connection to Cumulocity](./../tutorials/connect-c8y.md).

Reviewer 2: ```suggestion
Before starting anything, make sure [your device is connected to Cumulocity](./../tutorials/connect-c8y.md).
```

**Step 0**
Only if you didn't install `c8y_configuration_plugin` by the debian package,
you need one additional step to initialize the plugin. Run this command.

```shell
sudo c8y_configuration_plugin --init
```

**Step 1**

> Modify the `/etc/tedge/c8y/c8y-configuration-plugin.toml` file in the [format written in the specification](./../references/c8y-configuration-management.md#configuration).

Reviewer 2: ```suggestion
Add the files you want to configure from the cloud into the `/etc/tedge/c8y/c8y-configuration-plugin.toml` file using the [format described in the reference guide](./../references/c8y-configuration-management.md#configuration).
```

I would also:
* Add an example
* Say this file can be let empty and configured from the cloud
* Provide a link to the reference guide as a note after the example and not in the first sentence.

Author: Now `c8y_configuration_plugin --init` creates an example configuration file in `/etc/tedge/c8y/c8y-configuration-plugin.toml`. So it doesn't need to be empty.
